src/main/java/com/google/devtools/build/lib/remote/circuitbreaker/CircuitBreakerFactory.java

```diff
@@ -13,16 +13,32 @@
 // limitations under the License.
 package com.google.devtools.build.lib.remote.circuitbreaker;
 
-import com.google.common.collect.ImmutableSet;
 import com.google.devtools.build.lib.remote.Retrier;
 import com.google.devtools.build.lib.remote.common.CacheNotFoundException;
 import com.google.devtools.build.lib.remote.options.RemoteOptions;
+import io.grpc.Status;
+import java.util.function.Predicate;
+
+import static com.google.devtools.build.lib.remote.RemoteRetrier.fromException;
+
 
 /** Factory for {@link Retrier.CircuitBreaker} */
 public class CircuitBreakerFactory {
-
-  public static final ImmutableSet<Class<? extends Exception>> DEFAULT_IGNORED_ERRORS =
-      ImmutableSet.of(CacheNotFoundException.class);
+  public static final Predicate<? super Exception> DEFAULT_IGNORED_ERRORS =
```

Comment:
This is fundamentally the same as before. Since we already have the logic to check whether Bazel should retry on certain errors in `Retrier`, we want to reuse that instead of repeating the logic here.

Currently, we always [call](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/remote/Retrier.java;l=248;drc=5575ff24aff3c2ba6e487158f8d79f72d86e6800) `circuitBreaker.recordFailure(e)` if Bazel encountered an error. If we move the call after the [shouldRetry](https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/remote/Retrier.java;l=253;drc=5575ff24aff3c2ba6e487158f8d79f72d86e6800) check, we know errors are not ignored when they are arrived in `recordFailure`.

This makes CircuitBreaker no longer need to check whether it should ignore certain errors and makes it possible to use the CircuitBreaker for HTTP payload. 

Sorry if I didn't make it clear before.

Reply:
Thanks for the feedback.
You mean call recordFailure if error is not one of these `CANCELLED, UNKNOWN, DEADLINE_EXCEEDED, ABORTED, INTERNAL, UNAVAILABLE, RESOURCE_EXHAUSTED`.
 If I take diff with available grpc errors then leftover grpc errors are `INVALID_ARGUMENT`, `NOT_FOUND`, `ALREADY_EXISTS`, `PERMISSION_DENIED`, `UNAUTHENTICATED`, `FAILED_PRECONDITION` `OUT_OF_RANGE`, and `UNIMPLEMENTED` 

At highlevel it seems that left over errors can be treated as ignored error. But I believe being explicit could be better. And here I only want `NOT_FOUND` and `OUT_OF_RANGE` to marked as ignored errror.


Also  ignored errors are marked as success call for circuit breaker logic. [code](https://github.com/bazelbuild/bazel/pull/18539/files#diff-3437a1bb981f65cde4051a48e33811cbcc2b4094f4e0ae999060c52fa0303186R71-R89). The diff is not yet merged so I had to create pull request against master:HEAD.

I had two options then 
1. Add boolean argument `isErrorIgnored` as argument for recordFailure.
2.  Or call `recordSuccess` on ignored errors.


What do you say?

Reply:
Moved the logic to ignore failure to Retrier.